service/storage_service.cc

```diff
@@ -6119,6 +6155,180 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
     }
 }
 
+future<join_node_request_result> storage_service::join_node_request_handler(join_node_request_params params) {
+    if (!_raft_topology_change_enabled) {
+        throw std::runtime_error("The node is not configured to use topology over raft");
+    }
+
+    join_node_request_result result;
+    slogger.info("raft topology: received request to join from host_id: {}", params.host_id);
+
+    if (params.cluster_name != _db.local().get_config().cluster_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Cluster name check failed. This node cannot join the cluster "
+                                "because it expected cluster name \"{}\" and not \"{}\"",
+                                params.cluster_name,
+                                _db.local().get_config().cluster_name()),
+        };
+        co_return result;
+    }
+
+    if (params.snitch_name != _db.local().get_snitch_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Snitch name check failed. This node cannot join the cluster "
+                                "because it uses \"{}\" and not \"{}\"",
+                                params.snitch_name,
+                                _db.local().get_snitch_name()),
+        };
+        co_return result;
+    }
+
+    co_await _topology_state_machine.event.when([this] {
+        // The first node defines the cluster and inserts its entry to the
+        // `system.topology` without checking anything. It is unlikely but
+        // possible that the `join_node_request_handler` fires before the first
+        // node inserts its entry, therefore we might need to wait
+        // until that happens, here.
+        return !_topology_state_machine._topology.is_empty();
+    });
+
+    // TODO: Aborts?
```

Comment:
Well, you already have an `_abort_source` but just decided to ignore the exception for some reason.

Reply:
Now the abort requested exception is propagated.

Reply:
Should TODO be dropped then?

Reply:
Yeah, I removed it.